File: src/lib/photo-scoring.ts
import { getStop } from "./quests";
import type { BadgeTier, PhotoScore, VerificationResult } from "./types";
import { fetchLivePhotoScore, fetchLivePhotoVerification } from "./integrations/grabmaps";

export async function verifyPhoto(stopId: string, imageName: string): Promise<VerificationResult> {
  const live = await fetchLivePhotoVerification(stopId, imageName);
  if (live) return live;

  const stop = getStop(stopId);
  const normalizedName = imageName.toLowerCase();
  const expectedName = stop.demoPhotoName.toLowerCase();
  const passed =
    normalizedName === "demo-upload.jpg" ||
    normalizedName === expectedName ||
    normalizedName.includes(stop.id.replace(/^sg-|^bkk-/, "").slice(0, 6));

  return {
    passed,
    confidence: passed ? 0.92 : 0.31,
    reason: passed
      ? `Photo cues match ${stop.name}: signage, queue flow, and food-centre context are present.`
      : `The uploaded image does not contain enough visual evidence for ${stop.name}.`,
    source: "fallback"
  };
}

export async function scorePhoto(stopId: string, verification: VerificationResult): Promise<PhotoScore> {
  if (!verification.passed) {
    throw new Error("Photo must pass location verification before scoring.");
  }

  const live = await fetchLivePhotoScore(stopId, verification);
  if (live) return live;

  const stop = getStop(stopId);
  const score = stop.id === "sg-maxwell" ? 85 : 78;
  const tier = getTier(score);
  return {
    score,
    tier,
    categories: {
      locationMatch: 92,
      authenticity: 88,
      composition: score === 85 ? 82 : 74,
      culturalContext: 84,
      lighting: 79
    },
    feedback:
      score === 85
        ? "The queue in the foreground adds life; next time try catching a handwritten menu board to anchor the story even more."
        : "Strong place context. Step a little closer to include one local texture, like signage, a counter detail, or someone ordering.",
    unlockedLore: tier === "Gold" ? stop.bonusLore : "Score Gold to unlock the deeper local note for this stop.",
    source: "fallback"
  };
}

function getTier(score: number): Exclude<BadgeTier, "Locked"> {
  if (score >= 82) return "Gold";
  if (score >= 65) return "Silver";
  return "Bronze";
}

File: backend/app/photo_scoring.py
from app.data import get_stop
from app.models import PhotoScore, VerificationResult


async def verify_photo(stop_id: str, image_name: str) -> VerificationResult:
    stop = get_stop(stop_id)
    normalized_name = image_name.lower()
    expected_name = stop.demo_photo_name.lower()
    passed = (
        normalized_name == "demo-upload.jpg"
        or normalized_name == expected_name
        or stop.id.removeprefix("sg-").removeprefix("bkk-")[:6] in normalized_name
    )
    return VerificationResult(
        passed=passed,
        confidence=0.92 if passed else 0.31,
        reason=(
            f"Photo cues match {stop.name}: signage, queue flow, and food-centre context are present."
            if passed
            else f"The uploaded image does not contain enough visual evidence for {stop.name}."
        ),
        source="fallback",
    )


async def score_photo(stop_id: str, verification: VerificationResult) -> PhotoScore:
    if not verification.passed:
        raise ValueError("Photo must pass location verification before scoring.")

    stop = get_stop(stop_id)
    score = 85 if stop.id == "sg-maxwell" else 78
    tier = "Gold" if score >= 82 else "Silver" if score >= 65 else "Bronze"
    return PhotoScore(
        score=score,
        tier=tier,
        categories={
            "locationMatch": 92,
            "authenticity": 88,
            "composition": 82 if score == 85 else 74,
            "culturalContext": 84,
            "lighting": 79,
        },
        feedback=(
            "The queue in the foreground adds life; next time try catching a handwritten menu board to anchor the story even more."
            if score == 85
            else "Strong place context. Step a little closer to include one local texture, like signage, a counter detail, or someone ordering."
        ),
        unlockedLore=stop.bonus_lore if tier == "Gold" else "Score Gold to unlock the deeper local note for this stop.",
        source="fallback",
    )

